### PR TITLE
Fix: Exclude dropped chunks from the list of all chunks in the catalog

### LINKED ISSUE #206: Event streamer tries to read from dropped chunks
It seems to me that event streamer is trying to read from / work with dropped hypertable chunks, which of course it can't find.

```ini
CDC (Change Data Capture) for TimescaleDB Hypertable
timescaledb-event-streamer version 0.12.1-dev (git revision unknown; branch unknown)
Loading configuration file: config.toml
[2024/02/07T15:35:33.391] [INFO] [SystemCatalog] Selected tables for replication:  
[2024/02/07T15:35:33.391] [INFO] [SystemCatalog]   * [REDACTED] (type: Hypertable, replica identity: DEFAULT)  
[2024/02/07T15:35:33.391] [INFO] [Replicator] Discovered System Information:  
[2024/02/07T15:35:33.392] [INFO] [Replicator]   * PostgreSQL version 15.3  
[2024/02/07T15:35:33.392] [INFO] [Replicator]   * TimescaleDB version 2.10.3  
[2024/02/07T15:35:33.392] [INFO] [Replicator]   * PostgreSQL System Identity 7312038712204218402  
[2024/02/07T15:35:33.392] [INFO] [Replicator]   * PostgreSQL Timeline 13  
[2024/02/07T15:35:33.392] [INFO] [Replicator]   * PostgreSQL DatabaseName smartcitydb  
[2024/02/07T15:35:33.392] [INFO] [Replicator]   * PostgreSQL Types loaded 597  
[2024/02/07T15:35:33.392] [INFO] [FileStateStorage] Starting FileStateStorage at /tmp/statestorage.dat  
[2024/02/07T15:35:33.392] [INFO] [FileStateStorage] Loading FileStateStorage at /tmp/statestorage.dat  
[2024/02/07T15:35:35.285] [INFO] [ReplicationConnection] SystemId: 7312038712204218402, Timeline: 13, XLogPos: C0C/DAE6FC28, DatabaseName: [Redacted]  
[2024/02/07T15:35:37.294] [INFO] [ReplicationChannel] Reused replication slot: event_streamer_test  
[2024/02/07T15:35:38.154] [INFO] [FileStateStorage] Stopping FileStateStorage at /tmp/statestorage.dat  
[2024/02/07T15:35:38.154] [INFO] [FileStateStorage] Storing FileStateStorage at /tmp/statestorage.dat  
[2024/02/07T15:35:38.154] [INFO] [TaskDispatcher] TaskManager shutting down  
ERROR: relation "_timescaledb_internal._hyper_1_7_chunk" does not exist (SQLSTATE 42P01)
```

Checking chunks in timescaledb catalog:

```SQL
select * from _timescaledb_catalog.chunk where hypertable_id = 1;
```

| id | hypertable_id | schema_name           | table_name        | compressed_chunk_id | dropped | status | osm_chunk |
| -- | ------------- | --------------------- | ----------------- | ------------------- | ------- | ------ | --------- |
| 7  | 1             | _timescaledb_internal | _hyper_1_7_chunk  |                     | 1       | 0      | 0         |
| 8  | 1             | _timescaledb_internal | _hyper_1_8_chunk  |                     | 1       | 0      | 0         |
| 10 | 1             | _timescaledb_internal | _hyper_1_10_chunk |                     | 0       | 0      | 0         |
| 11 | 1             | _timescaledb_internal | _hyper_1_11_chunk |                     | 0       | 0      | 0         |

Checking the tables in `_timescaledb_internal` schema:

```SQL
SELECT table_name FROM information_schema.tables WHERE table_schema='_timescaledb_internal' AND table_type='BASE TABLE' order by table_name asc;
```

| table_name
| --
| _hyper_1_10_chunk
| _hyper_1_11_chunk
| _hyper_2_2_chunk
| _hyper_2_9_chunk
| _materialized_hypertable_2
| bgw_job_stat
| bgw_policy_chunk_stats
| job_errors

My best idea for fixing this would be to:
- either not return dropped chunks in the selected hypertable when `GetAllChunks()` is called [here](https://github.com/noctarius/timescaledb-event-streamer/blob/main/internal/systemcatalog/systemcatalog.go#L378)
- filter the list of returned chunks somewhere around [here](https://github.com/noctarius/timescaledb-event-streamer/blob/main/internal/replication/replicator.go#L394-L403), because we're filtering them anyway at that point

does either of these make sense?

I've used the following patch and it worked, but I'm wondering if this is the right thing to do.

```patch
diff --git a/internal/systemcatalog/systemcatalog.go b/internal/systemcatalog/systemcatalog.go
index 6aea5ac..1655449 100644
--- a/internal/systemcatalog/systemcatalog.go
+++ b/internal/systemcatalog/systemcatalog.go
@@ -375,7 +375,9 @@ func (sc *systemCatalog) GetAllChunks() []systemcatalog.SystemEntity {
 	chunkTables := make([]systemcatalog.SystemEntity, 0)
 	for _, chunk := range sc.chunks {
 		if sc.IsHypertableSelectedForReplication(chunk.HypertableId()) {
-			chunkTables = append(chunkTables, chunk)
+			if !chunk.Dropped() {
+				chunkTables = append(chunkTables, chunk)
+			}
 		}
 	}
 	return chunkTables
```

### PR DESCRIPTION
**Pull request by vidosits**
Fixes #206

Unit tests have to be written, but internal/systemcatalog/systemcatalog.go is missing its tests altogether and I'm kinda lost where to start so as to not to just test basic stuff, but @noctarius the gentleman he is has offered to add unit tests

### CODE REVIEW

**Comment by vidosits** [comment]
hey, @noctarius could you check if this is the kind of test you were thinking of?

**Comment by noctarius** [comment]
Looks like a real good initial test. Have you made sure the test fails without your fix? I'm not sure that this case fixes your specific issue. Happy to test it though :-)

**Comment by noctarius** [comment]
It looks like it's working without the additional check. I wonder when dropped is set to true, over the chunk being dropped immediately 🤔

Let me investigate the Timescale sourcecode. What version do you use? Maybe it's a recent change?!

**Comment by vidosits** [comment]
you're right of course, this test is a bait in it's current form, it passes without the committed "fix" either way. I think the reason is a combination of the following:

The test itself does not touch [this](https://github.com/noctarius/timescaledb-event-streamer/blob/main/internal/systemcatalog/systemcatalog.go#L374) code path. So even if the test failed then the fix wouldn't fix it. The reason is because we're not collecting chunks using `func (sc *systemCatalog) GetAllChunks()`, but rather with an SQL query on the timescale internal schema.

During this test we're not actually testing the internals, we're just basically testing if we successfully dropped the chunks. It's a bad test.

The whole reason I needed this fix is because of the following scenario:

- We have a retention policy on one of our tables
- The table is huge and gets a lot of queries per minute
- I *think* (not quite sure) but timescale can't get a [lock on the chunk](https://docs.timescale.com/use-timescale/latest/data-retention/troubleshooting/#dropping-chunks-times-out) because of a continuous aggregate or some other query
- I'm actually unsure about the above because the chunk *does* get dropped, but then metadata stays in the internal chunks table, but it correctly gets marked as "dropped" see table from #206

Also the test is wrong because of the above: when I'm dropping chunks with `SELECT drop_chunks()` in the test they all get dropped and I think there's no corresponding row left in the mock db indicating that the chunk was there but with dropped = true column set

To summarize:
- thanks for taking a look at the test, it's very obviously wrong in multiple places
- we'd have to somehow simulate (manually setting dropped=True with SQL?) some chunks being marked as dropped in the test
- we'd have to adjust the test so that it actually touches the function (`func (sc *systemCatalog) GetAllChunks()`) modified by the fix commit

**Comment by vidosits** [comment]
> It looks like it's working without the additional check. I wonder when dropped is set to true, over the chunk being dropped immediately 🤔
> 
> Let me investigate the Timescale sourcecode. What version do you use? Maybe it's a recent change?!

We're using `timescale/timescaledb-ha:pg15-ts2.10` image with the v0.33.1 timescaledb-single helm chart from charts.timescale.com

**Comment by vidosits** [comment]
No problem, thanks for taking the time and looking at my silly problem

**Comment by noctarius** [comment]
Not silly at all. It's a valid issue and should be fixed. I guess, I'll probably fall back, forcing the dropped field to true 😅

It may be possible, that it is set first, in it's own transaction, and then the actual drop happens, to prevent the data being "available" for large chunks when actually dropping may take a second or two.

**Comment by noctarius** [comment]
Fixed via #305
